api/dao/hierarchy.py

```diff
@@ -319,53 +319,59 @@ def dict_fileinfos(infos):
-    acquisition['modified'] = now
-    acquisition_obj = _update_container({'_id': acquisition_id}, acquisition, 'acquisitions')
-    if acquisition_obj is None:
-        raise APIStorageException('acquisition doesn''t exist')
-    if acquisition.get('timestamp'):
-        session_obj = config.db.sessions.find_one_and_update(
-            {'_id': acquisition_obj['session']},
-            {
-                '$min': dict(timestamp=acquisition['timestamp']),
-                '$set': dict(timezone=acquisition.get('timezone'))
-            },
-            return_document=pymongo.collection.ReturnDocument.AFTER
-        )
-        config.db.projects.find_one_and_update(
-            {'_id': session_obj['project']},
-            {
-                '$max': dict(timestamp=acquisition['timestamp']),
-                '$set': dict(timezone=acquisition.get('timezone'))
-            }
-        )
-    session_obj = None
-    if session:
-        session['modified'] = now
-        session_obj = _update_container({'_id': acquisition_obj['session']}, session, 'sessions')
-    if project:
-        project['modified'] = now
-        if not session_obj:
-            session_obj = config.db.sessions.find_one({'_id': acquisition_obj['session']})
-        _update_container({'_id': session_obj['project']}, project, 'projects')
-    return acquisition_obj
-
-def _update_container(query, update, cont_name):
-    return config.db[cont_name].find_one_and_update(
-        query,
-        {
-            '$set': util.mongo_dict(update)
-        },
+    if c_metadata.get('timestamp'):
+        c_metadata['timestamp'] = dateutil.parser.parse(c_metadata['timestamp'])
+    c_metadata['modified'] = now
+    c_obj = _update_container({'_id': cid}, {}, c_metadata, container_type)
+    if c_obj is None:
+        raise APIStorageException('container does not exist')
+    if container_type in ['session', 'acquisition']:
+        update_timestamp = True if c_metadata.get('timestamp') else False
+        _update_hierarchy(c_obj, container_type, metadata, update_timestamp)
+    return c_obj
+
+
+def _update_container(query, update, set_update, container_type):
+    coll_name = container_type if container_type.endswith('s') else container_type+'s'
+    update['$set'] = util.mongo_dict(set_update)
+    return config.db[coll_name].find_one_and_update(query,update,
         return_document=pymongo.collection.ReturnDocument.AFTER
     )
 
+
+def _update_hierarchy(container, container_type, metadata, update_timestamp=False):
+    project_id = container.get('project') # for sessions
+    now = datetime.datetime.utcnow()
+
+    if container_type == 'acquisition':
+        update = {}
+        session = metadata.get('session', {})
+        session_obj = None
+        if update_timestamp:
+            update['$min'] = dict(timestamp=container['timestamp'])
+            session['timezone'] = dict(timezone=container.get('timezone'))
```

Comment:
After offline conversation, we conclude that the engine placer will only do as it is told and not have side effects, such as updating session timestamps without being explicitly told to do so. Reaper uploads are a different story, though. Side effect will remain there.